src/shared/bpf-program.c

```diff
@@ -76,6 +76,10 @@ int bpf_program_load_kernel(BPFProgram *p, char *log_buf, size_t log_size) {
                 return 0;
         }
 
+        /* FIXME: Clang doesn't 0-pad with structured initialization, causing
+         * the kernel to reject the bpf_attr as invalid. Remove memsets once
```

Comment:
C standard doesn't dictate that padding is zero-initialized, so technically it's not a compiler bug :)

Reply:
True, but it looks like GCC does it, as we've been shipping with this code for a while and nobody complained so far. I don't believe any distro builds systemd with Clang, although I do not know for sure.

Reply:
> nobody complained so far

[Back in 2018](https://github.com/systemd/systemd/issues/9666#issuecomment-449681984) the test built with clang didn't fail loudly (it was just skipped) so it seems it was probably harder to notice.

Reply:
Interesting, do you still have a repro for that? If so, does this change fix it?

Reply:
question is why the kernel refuses non-initialized padding in the first place...

Reply:
It looks intentional:

https://github.com/torvalds/linux/blob/v5.9/kernel/bpf/syscall.c#L65

```
 * If we're handed a bigger struct than we know of, ensure all the unknown bits
 * are 0 - i.e. new user-space does not rely on any kernel feature extensions
 * we don't know about yet.
```

Reply:
> do you still have a repro for that?

It should be possible to just build `systemd` manually with `clang`

> If so, does this change fix it?

No, it doesn't. With this patch applied the test built with `gcc` passes on my machine. With `clang` it is still skipped.

```
$ clang --version
clang version 10.0.1 (Fedora 10.0.1-3.fc32)

$ cat /proc/version
Linux version 5.9.16-100.fc32.x86_64 (mockbuild@bkernel02.iad2.fedoraproject.org) (gcc (GCC) 10.2.1 20201125 (Red Hat 10.2.1-9), GNU ld version 2.34-6.fc32) #1 SMP Mon Dec 21 14:10:00 UTC 2020

$ systemd-detect-virt
oracle

$ sudo ./build/test-bpf-firewall
Found container virtualization none.
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/dbus/system_bus_socket...
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.34 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=StartTransientUnit cookie=3 reply_cookie=0 signature=ssa(sv)a(sa(sv)) error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.34 path=n/a interface=n/a member=n/a cookie=498 reply_cookie=3 signature=o error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.DBus destination=:1.34 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=4294967295 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.34 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=2 signature= error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.systemd1',path='/org/freedesktop/systemd1',interface='org.freedesktop.systemd1.Manager',member='JobRemoved' successfully installed.
Got message type=signal sender=:1.1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=503 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
Got result done/Success for job test-bpf-firewall-5e1ac56829dd2981.scope
Bus n/a: changing state RUNNING → CLOSED
Found cgroup2 on /sys/fs/cgroup/unified, unified hierarchy for systemd controller
Can't load kernel CGROUP SKB BPF program, BPF firewalling is not supported: Invalid argument
test-bpf-firewall: BPF firewalling not supported, skipping tests.
```

It seems this PR fixes something else

Reply:
Strange, the error it fixes is later than what you are getting. So in your case it stops earlier. You can see the log here: https://github.com/systemd/systemd/issues/18164

Reply:
On my machine, Debian testing, the test works just fine when building with clang 11.

Reply:
Could you show the first bpf call? With clang11 I got the same error:
```
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_CGROUP_SKB, insn_cnt=2, insns=[{code=BPF_ALU64|BPF_K|BPF_MOV, dst_reg=BPF_REG_0, src_reg=BPF_REG_0, off=0, imm=0x1}, {code=BPF_JMP|BPF_K|BPF_EXIT, dst_reg=BPF_REG_0, src_reg=BPF_REG_0, off=0, imm=0}], license="GPL", log_level=0, log_size=0, log_buf=NULL, kern_version=KERNEL_VERSION(0, 0, 0), prog_flags=0, prog_name="", prog_ifindex=0, expected_attach_type=BPF_CGROUP_INET_INGRESS, prog_btf_fd=0, func_info_rec_size=0, func_info=NULL, func_info_cnt=0, line_info_rec_size=0, line_info=NULL, line_info_cnt=0, attach_btf_id=0, attach_prog_fd=0, /* bytes 116..119 */ "\x5e\x7f\x00\x00"}, 120) = -1 EINVAL (Invalid argument)
```

Reply:
```
$ sudo buildclang/test-bpf-firewall
Found container virtualization none.
<...>
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Got EBADF when using BPF_F_ALLOW_MULTI, which indicates it is supported. Yay!
BPF firewalling with BPF_F_ALLOW_MULTI supported. Yay!
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Unified cgroup hierarchy is located at /sys/fs/cgroup/system.slice/test-bpf-firewall-bda7de916d715b6.scope/3138a86134f9ccce.
Controller 'cpu' supported: yes
<...>
```

Reply:
Thank you! Could you run it under `strace` to get the first bpf call (where it returns EBADF)? Also, I'm wondering what else apart from CC=clang CXX=clang did you pass?

Reply:
CC=clang CXX=clang++ meson buildclang --optimization=1

Reply:
```
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_CGROUP_SKB, insn_cnt=2, insns=0x15cbbc0, license="GPL", log_level=0, log_size=0, log_buf=NULL, kern_version=KERNEL_VERSION(0, 0, 0), prog_flags=0, prog_name="", prog_ifindex=0, expected_attach_type=BPF_CGROUP_INET_INGRESS, prog_btf_fd=0, func_info_rec_size=0, func_info=NULL, func_info_cnt=0, line_info_rec_size=0, line_info=NULL, line_info_cnt=0, attach_btf_id=0, attach_prog_fd=0}, 120) = 3
bpf(BPF_PROG_DETACH, {target_fd=-1, attach_type=BPF_CGROUP_INET_EGRESS}, 120) = -1 EBADF (Bad file descriptor)
bpf(BPF_PROG_ATTACH, {target_fd=-1, attach_bpf_fd=-1, attach_type=BPF_CGROUP_INET_EGRESS, attach_flags=BPF_F_ALLOW_MULTI, replace_bpf_fd=0}, 120) = -1 EBADF (Bad file descriptor)
Got EBADF when using BPF_F_ALLOW_MULTI, which indicates it is supported. Yay!
BPF firewalling with BPF_F_ALLOW_MULTI supported. Yay!
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_CGROUP_SKB, insn_cnt=2, insns=0x15ce860, license="GPL", log_level=1, log_size=65535, log_buf="", kern_version=KERNEL_VERSION(0, 0, 0), prog_flags=0, prog_name="", prog_ifindex=0, expected_attach_type=BPF_CGROUP_INET_INGRESS, prog_btf_fd=0, func_info_rec_size=0, func_info=NULL, func_info_cnt=0, line_info_rec_size=0, line_info=NULL, line_info_cnt=0, attach_btf_id=0, attach_prog_fd=0}, 120) = 3
bpf(BPF_OBJ_PIN, {pathname="/sys/fs/bpf/test-dropper", bpf_fd=3, file_flags=0}, 120) = 0
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Unified cgroup hierarchy is located at /sys/fs/cgroup/system.slice/test-bpf-firewall-7b19727bf6658ff0.scope/2cf62dbf35a69f1.
bpf(BPF_PROG_LOAD, {prog_type=BPF_PROG_TYPE_CGROUP_DEVICE, insn_cnt=2, insns=0x15ce0d0, license="GPL", log_level=0, log_size=0, log_buf=NULL, kern_version=KERNEL_VERSION(0, 0, 0), prog_flags=0, prog_name="", prog_ifindex=0, expected_attach_type=BPF_CGROUP_INET_INGRESS, prog_btf_fd=0, func_info_rec_size=0, func_info=NULL, func_info_cnt=0, line_info_rec_size=0, line_info=NULL, line_info_cnt=0, attach_btf_id=0, attach_prog_fd=0}, 120) = 7
Controller 'cpu' supported: yes
```

Reply:
I've just spun up a Debian Testing VM. Looks like there it's 100% reproducible without `--optimization=1`. `CC=clang CXX=clang++ meson buildclang` should suffice I think

Reply:
right, --optimization=1 is cargo-culted from the CI build scripts - no idea why it was first added

Reply:
Judging by https://github.com/systemd/systemd/pull/18185 it's skipped on Ubuntu as well without `--optimization`. Looks like the bpf part is still incompatible with `clang`.

Reply:
Looks like they are definitely different issues then - strange stuff

Reply:
Given that the test can get past that point with the following patch applied:
```diff
diff --git a/src/shared/bpf-program.c b/src/shared/bpf-program.c
index 9da8c109b8..ab72985d91 100644
--- a/src/shared/bpf-program.c
+++ b/src/shared/bpf-program.c
@@ -81,15 +81,13 @@ int bpf_program_load_kernel(BPFProgram *p, char *log_buf, size_t log_size) {
         // https://github.com/torvalds/linux/blob/v5.9/kernel/bpf/syscall.c#L65
         // Ideally it should behave like GCC, so that we can remove these workarounds.
         zero(attr);
-        attr = (union bpf_attr) {
-                .prog_type = p->prog_type,
-                .insns = PTR_TO_UINT64(p->instructions),
-                .insn_cnt = p->n_instructions,
-                .license = PTR_TO_UINT64("GPL"),
-                .log_buf = PTR_TO_UINT64(log_buf),
-                .log_level = !!log_buf,
-                .log_size = log_size,
-        };
+        attr.prog_type = p->prog_type;
+        attr.insns = PTR_TO_UINT64(p->instructions);
+        attr.insn_cnt = p->n_instructions;
+        attr.license = PTR_TO_UINT64("GPL");
+        attr.log_buf = PTR_TO_UINT64(log_buf);
+        attr.log_level = !!log_buf;
+        attr.log_size = log_size;

         p->kernel_fd = bpf(BPF_PROG_LOAD, &attr, sizeof(attr));
         if (p->kernel_fd < 0)
```
I'd say it's the same issue.

Reply:
For different points of execution though? The CI failed without this PR, and now passes - all the rest being the same. There was an assert hit at some point

Reply:
that said, if we have to forego structured init then let's just do it - could you send a PR to that effect?

Reply:
I'm not sure how to prepare the CI for PRs like this. As long as `systemd` builds with just one optimization level I don't think we can guarantee that it works properly. Apart from that, even if I tweaked the build pipeline (where `systemd` is built with -02, -03, -lto and so on) it seems I couldn't specify that `test-bpf-firewall` shouldn't be skipped there. I mean, I have no problem with one-shot patches but given that, as far as I know, the bpf part is going to be changed soon it would probably be better to figure out how to cover it.

Reply:
I just meant something like this: https://github.com/systemd/systemd/pull/18195 - ie, remove the opt build flag and change the initialization style. If the test doesn't fail anymore (it would without --optimization) then we should be somewhat good